apollo-router/src/uplink/mod.rs

```diff
@@ -1,6 +1,690 @@
+                    id: last_id.clone(),
+                }
+                .into(),
+            );
+
+            match fetch::<Query>(&query_body, &mut endpoints.iter(), timeout).await {
+                Ok(value) => {
+                    let response: UplinkResponse<Response> = value.into();
+                    match response {
+                        UplinkResponse::Result {
+                            id,
+                            response,
+                            delay,
+                        } => {
+                            last_id = Some(id);
+                            if sender.send(Ok(response)).await.is_err() {
+                                break;
+                            }
+
+                            interval = Duration::from_secs(delay);
+                        }
+                        UplinkResponse::Unchanged { id, delay } => {
+                            tracing::trace!("uplink response did not change");
+                            // Preserve behavior for schema uplink errors where id and delay are not reset if they are not provided on error.
+                            if let Some(id) = id {
+                                last_id = Some(id);
+                            }
+                            if let Some(delay) = delay {
+                                interval = Duration::from_secs(delay);
+                            }
+                        }
+                        UplinkResponse::Error {
+                            retry_later,
+                            message,
+                            code,
+                        } => {
+                            let err = if retry_later {
+                                Err(Error::UplinkError { code, message })
+                            } else {
+                                Err(Error::UplinkErrorNoRetry { code, message })
+                            };
+                            if sender.send(err).await.is_err() {
+                                break;
+                            }
+                            if !retry_later {
+                                break;
+                            }
+                        }
+                    }
+                }
+                Err(err) => {
+                    if sender.send(Err(err)).await.is_err() {
+                        break;
+                    }
+                }
+            }
+
+            tokio::time::sleep(interval).await;
+        }
+    };
+    drop(tokio::task::spawn(task.with_current_subscriber()));
+
+    ReceiverStream::new(receiver)
+}
+
+pub(crate) async fn fetch<Query>(
+    request_body: &QueryBody<Query::Variables>,
+    urls: &mut impl Iterator<Item = &Url>,
+    timeout: Duration,
+) -> Result<Query::ResponseData, Error>
+where
+    Query: graphql_client::GraphQLQuery,
+{
+    for url in urls {
+        return match http_request::<Query>(url.as_str(), request_body, timeout).await {
```

Comment:
I think something is wrong here... I think we're not iterating over the urls, the return should be inside a match arm. Could you check the behavior ?

Reply:
https://github.com/apollographql/router/pull/2537/files/702211513aa03e6f54b983d37e625ca2dbdba036#diff-b53d60efefb7c095c121333874abdae658614f2538f8d41ef50e83e03c9542fcR219 will cause it to not return.
I'll change it though as I think your suggestion is clearer.

Reply:
[0646d94](https://github.com/apollographql/router/pull/2537/commits/0646d94e09a75800f58f899424a1d4aa4c946558)